# Written Assignment 05

*Author: Logan Reine*

### Headings

In [1]:
%run MLequations_v2.ipynb
import warnings
warnings.simplefilter(action = 'ignore', category = FutureWarning)

### Data Files

In [2]:
corrupt = pd.read_csv("corruption-data.csv")
surf = pd.read_csv("surf.tsv", delimiter = '\t')
email = pd.read_csv('spam_email.tsv', delimiter = '\t')
rec_sys = pd.read_csv('rec_system.tsv', delimiter = '\t')
houses = pd.read_csv('house_prices.tsv', delimiter = '\t')

# **1 A Nearest Neighbor Machine Learning Algorithm**

The table below lists a dataset that was used to create a nearest neighbor model that predicts whether it will be a good day to go surfing.

In [3]:
surf

ID  Wave Size (ft)  Wave Period (secs)  Wind Speed (MPH/hr) Good to Surf
0   1               6                  15                    5          yes
1   2               1                   6                    9           no
2   3               7                  10                    4          yes
3   4               7                  12                    3          yes
4   5               2                   2                   10           no
5   6              10                   2                   20           no

Assuming that the model uses Euclidean distance to find the nearest neighbor, what prediction will the model return for each of the following query instances:

**q1 = {'Wave Size (ft)': 8, 'Wave Period (secs)': 15, 'Wind Speed (MPH/hr)': 2}**  
**q2 = {'Wave Size (ft)': 8, 'Wave Period (secs)': 2, 'Wind Speed (MPH/hr)': 18}**  
**q3 = {'Wave Size (ft)': 6, 'Wave Period (secs)': 11, 'Wind Speed (MPH/hr)': 4}**  

In [4]:
q1 = {'Wave Size (ft)': 8, 'Wave Period (secs)': 15, 'Wind Speed (MPH/hr)': 2}
q2 = {'Wave Size (ft)': 8, 'Wave Period (secs)': 2, 'Wind Speed (MPH/hr)': 18}
q3 = {'Wave Size (ft)': 6, 'Wave Period (secs)': 11, 'Wind Speed (MPH/hr)': 4}

surf_c = format_frame(surf, surf.iloc[0, -1])

n = e_dist(q1, surf_c)
form('Euclidean Distance', n, 1, maj_vote(n, surf_c))

This 1-NN Euclidean Distance model returns: 

(INDEX, VALUE)
(3, 3.3166247903554)

The 1 nearest neighbors predict a target feature of: True 


In [5]:
n = e_dist(q2, surf_c)
form('Euclidean Distance', n, 1, maj_vote(n, surf_c)) 

This 1-NN Euclidean Distance model returns: 

(INDEX, VALUE)
(5, 2.8284271247461903)

The 1 nearest neighbors predict a target feature of: False 


In [6]:
n = e_dist(q3, surf_c)
form('Euclidean Distance', n, 1, maj_vote(n, surf_c)) 

This 1-NN Euclidean Distance model returns: 

(INDEX, VALUE)
(2, 1.4142135623730951)

The 1 nearest neighbors predict a target feature of: True 


### Q1 Good to Surf = *yes*  
### Q2 Good to Surf = *no*  
### Q3 Good to Surf = *yes*  

# **2 Email Spam Filtering**  



Email spam filtering models often use a bag-of-words representation for emails. In a bag-of-words representation, the descriptive features that describe a document (in our case, an email) each represent how many times a particular word occurs in the document. One descriptive feature is included for each word in a predefined dictionary. The dictionary is typically defined as the complete set of words that occur in the training dataset.  

The table below lists the bag-of-words representation for the following five emails and a target feature, spam, whether they are spam emails or genuine emails:  

  1. money, money, money
  2. free money of free gambling fun
  3. gambling of fun
  4. machine learning of fun, fun, fun
  5. free machine learning

In [7]:
email

ID  money  free  of  gambling  fun  machine  learning   spam
0   1      3     0   0         0    0        0         0   True
1   2      1     2   1         1    1        0         0   True
2   3      0     0   1         1    1        0         0   True
3   4      0     0   1         0    3        1         1  False
4   5      0     1   0         0    0        1         1  False

1. What target level would a nearest neighbor model using Euclidean distance return for the following email: **machine learning of free**?

In [8]:
q1 = pd.DataFrame({
    'money': [0],
    'free': [1],
    'of': [1], 
    'gambling': [0],
    'fun': [0],
    'machine': [1],
    'learning': [1]
})

n = e_dist(q1, email)
form('Euclidean Distance', n, 1, maj_vote(n, email))  

This 1-NN Euclidean Distance model returns: 

(INDEX, VALUE)
(4, 1.0)

The 1 nearest neighbors predict a target feature of: False 


  2. What target level would a k ⁣− ⁣NN model with k = 3 and using the **Euclidean distance** return for the same query?

In [9]:
n = e_dist(q1, email, 3)
form('Euclidean Distance', n, 3, maj_vote(n, email))  

This 3-NN Euclidean Distance model returns: 

(INDEX, VALUE)
(4, 1.0)
(2, 2.23606797749979)
(1, 2.449489742783178)

The 3 nearest neighbors predict a target feature of: True 


  3. What target level would a weighted k ⁣− ⁣NN model with k=5 and using a weighting scheme of the reciprocal of the squared Euclidean distance between the neighbor and the query, return for the query?

In [10]:
print('A 5-NN Weighted Euclidean Distance would return the target level:', w_knn(q1, email, 5))

True Weight: 0.44358974358974357 
False Weight: 1.1
A 5-NN Weighted Euclidean Distance would return the target level: False


  4. What target level would a k ⁣− ⁣NN model with k = 3 and using the Manhattan distance return for the same query?

In [11]:
n = man_d(q1, email, 3)
form('Manhattan distance', n, 3, maj_vote(n, email)) 

This 3-NN Manhattan distance model returns: 

(INDEX, VALUE)
(4, 1.0)
(3, 4.0)
(2, 5.0)

The 3 nearest neighbors predict a target feature of: False 


  5. There are a lot of zero entries in the spam bag-of-words dataset. This is indicative of sparse data and is typical for text analytics. Cosine similarity is often a good choice when dealing with sparse non-binary data. What target level would a 3−NN model using the cosine similarity return for the query?

In [12]:
n = cos_sim(q1, email, 3)
form('Cosine Similarity', n, 3, maj_vote(n, email))                   

This 3-NN Cosine Similarity model returns: 

(INDEX, VALUE)
(3, 0.4330127018922194)
(1, 0.5303300858899106)
(4, 0.8660254037844388)

The 3 nearest neighbors predict a target feature of: False 


# **3 Corruption Prediction**  


The predictive task in this question is to predict the level of corruption in a country based on a range of macro-economic and social features. The following table lists some countries described by the following descriptive features:  

  1. **Life Exp** - the mean life expectancy at birth
  2. **Top-10 Income** – the percentage of the annual income of the country that goes to the top 10% of earners
  3. **Infant Mor** – the number of infant deaths per 1,000 births
  4. **Mil Spend** – the percentage of GDP spent on the military
  5. **School Years** – the mean number years spent in school by adult females

The target feature is the **Corruption Perception Index** (CPI). The **CPI** measures the perceived levels of corruption in the public sector of countries and ranges from 0 (highly corrupt) to 100 (very clean).

In [13]:
corrupt

ID    Exp  Income   Mor  Spending  Years     CPI
0   Afghanistan  59.61   23.21  74.3      4.44    0.4  1.5171
1         Haiti  45.00   47.67  73.1      0.09    3.4  1.7999
2       Nigeria  51.30   38.23  82.6      1.07    4.1  2.4493
3         Egypt  70.48   26.58  19.6      1.86    5.3  2.8622
4     Argentina  75.77   32.30  13.3      0.76   10.1  2.9961
5         China  74.87   29.98  13.7      1.95    6.4  3.6356
6        Brazil  73.12   42.93  14.5      1.43    7.2  3.7741
7        Israel  81.30   28.80   3.6      6.77   12.5  5.8069
8         U.S.A  78.51   29.85   6.3      4.72   13.7  7.1357
9       Ireland  80.15   27.23   3.5      0.60   11.5  7.5360
10         U.K.  80.09   28.49   4.4      2.59   13.0  7.7751
11      Germany  80.24   22.07   3.5      1.31   12.0  8.0461
12       Canada  80.99   24.79   4.9      1.42   14.2  8.6725
13    Australia  82.09   25.40   4.2      1.86   11.5  8.8442
14       Sweden  81.43   22.18   2.4      1.27   12.8  9.2985
15  New Zealand  80.67   27.81   4.9      1.13   12.3  9.4627

We will use Russia as our query country for CPI prediction with the following values for the descriptive features:  

    Life Expectancy = 67.62
    Top-10 Income = 31.68
    Infant Mortality = 10.0
    Military Spending = 3.87
    School Years = 12.9

  a. What value would a 3-nearest neighbor prediction model using Euclidean distance return for the CPI of Russia?

In [14]:
russia = pd.DataFrame({
    'ID': ['Russia'],
    'Exp': [67.62],
    'Income': [31.68],
    'Mor': [10.],
    'Spending': [3.87],
    'Years': [12.9],
    'CPI': [None]
})

n = e_dist(russia, corrupt, 3)
form('Euclidean Distance', n, 3, maj_vote_f(n, corrupt), 'CPI')

This 3-NN Euclidean Distance model returns: 

(INDEX, VALUE)
(4, 9.757509928255255)
(5, 10.72748339546606)
(8, 11.704422241187302)

The 3 nearest neighbors predict a target feature of: 4.589133333333334 CPI


  b. What value would a weighted k ⁣− ⁣NN prediction model return for the CPI of Russia? Use k=16 (i.e., the full dataset) and a weighting scheme of the reciprocal of the squared Euclidean distance between the neighbor and the query.

In [15]:
wn = w_knn_f(russia, corrupt, 16)
print('This 16-NN Weighted Euclidean Distance model would predict:', wn, 'CPI')

This 16-NN Weighted Euclidean Distance model would predict: 5.908707536841123 CPI


c. The descriptive features in this dataset are of different types. For example, some are percentages, others are measured in years, and others are measured in counts per 1,000. We should always consider normalizing our data, but it is particularly important to do this when the descriptive features are measured in different units. What value would a 3-nearest neighbor prediction model using Euclidean distance return for the CPI of Russia when the descriptive features have been normalized using range normalization?

In [16]:
# Concatenate query and dataframe
query_frame = pd.concat([corrupt, russia], ignore_index = True)

# normalize query and dataframe together
norm_query_frame = norm(query_frame)

# separate query and frame
norm_russia = norm_query_frame.iloc[16:, 0:]

# adjust query index
norm_russia = norm_russia.rename(index = {16: 0})

# normalized dataframe without russia
norm_corrupt = norm_query_frame.drop(16)

n = e_dist(norm_russia, norm_corrupt, 3)
form('Euclidean Distance', n, 3, maj_vote_f(n, norm_corrupt), 'CPI')

This 3-NN Euclidean Distance model returns: 

(INDEX, VALUE)
(8, 0.336150849995879)
(10, 0.4125642049866904)
(4, 0.5554148303798639)

The 3 nearest neighbors predict a target feature of: 5.968966666666667 CPI


d. What value would a weighted k ⁣− ⁣NN prediction model—with k=16 (i.e., the full dataset) and using a weighting scheme of the reciprocal of the squared Euclidean distance between the neighbor and the query—return for the CPI of Russia when it is applied to the range-normalized data?

In [17]:
wn = w_knn_f(norm_russia, norm_corrupt, 16)
print('This 16-NN Weighted Euclidean Distance model would predict:', wn, 'CPI')

This 16-NN Weighted Euclidean Distance model would predict: 6.63466119877024 CPI


e. The actual 2011 CPI for Russia was 2.4488. Which of the predictions made was the most accurate? Why do you think this was?

*Well, with intuition alone I would assume a weighted and normalized set would predict the nearest neighbor, but my answer for 3.a is the closest to the 2011 actual with a CPI of 4.5891.  This made me think my functions/calculations might've been off, so I went back and re-calculated earlier problems, but the equations/results were accurate.  I do not know how to logically declare why my answer for 3.a was the closest, as requested by this question.  I know there is always a possibility for an instance query to return a result that's not an accurate representation of the data based on a "freak" instance.  But I don't know where my equations are getting mixed up arithmetically.  I've actually delayed submitting this assignment by several days so I could troubleshoot my code just to make sure, and to no avail.  So reluctantly, I'll 

# 4 Recommender systems

You have been given the job of building a recommender system for a large online shop that has a stock of over 100,000 items. In this domain the behavior of customers is captured in terms of what items they have bought or not bought. For example, the following table lists the behavior of two customers in this domain for a subset of the items that at least one of the customers has bought.

In [18]:
rec_sys

ID  Item 107  Item 498  Item 7256  Item 28063  Item 75328
0   1      True      True       True       False       False
1   2      True     False      False        True        True

The company has decided to use a similarity-based model to implement the recommender system.

  1. Which of the following three similarity indexes do you think the system should be based on?

*I think Russel-Rao would do well in this context.  If the online store has over 100,000 items, then that's a lot of inventory a hypothetical shopper might not have even seen.  The Russel-Rao index doesn't account for any co-absences, so any obscure items not viewed by a shopper wouldn't affect the formula.  But instead the Russel-Rao index would almost convey a ratio of (total purchases) / (total items viewed).  It would assosciate similar purchases with other buyers, versus base suggestions on what buyers didn't buy*

  2. What items will the system recommend to the following customer? Assume that the recommender system uses the similarity index you chose in the first part of this question and is trained on the sample dataset listed above. Also assume that the system generates recommendations for query customers by finding the customer most similar to them in the dataset and then recommending the items that this similar customer has bought but that the query customer has not bought.

In [19]:
query = pd.DataFrame({
    'ID': ['Query'],
    'Item 107': [True],
    'Item 498': [False],
    'Item 7256': [True],
    'Item 28063': [False],
    'Item 75328': [False]
})
query

ID  Item 107  Item 498  Item 7256  Item 28063  Item 75328
0  Query      True     False       True       False       False

In [20]:
q1 = rr(query.iloc[0:,1:], rec_sys.iloc[[0],1:])
q2 = rr(query.iloc[0:,1:], rec_sys.iloc[[1],1:])

print('\nRussel-Rao Index\nQuery 1:', q1, '\nQuery 2:', q2)


QUERY
    Item 107  Item 498  Item 7256  Item 28063  Item 75328
0      True     False       True       False       False

DATA INSTANCE
    Item 107  Item 498  Item 7256  Item 28063  Item 75328
0      True      True       True       False       False

QUERY
    Item 107  Item 498  Item 7256  Item 28063  Item 75328
0      True     False       True       False       False

DATA INSTANCE
    Item 107  Item 498  Item 7256  Item 28063  Item 75328
1      True     False      False        True        True

Russel-Rao Index
Query 1: 0.4 
Query 2: 0.2


*Since Query 1 has a higher similarity index, it would be more suitable for recommendations.  Next, the recommender system would check where values were true in the data instance and false in the query instance.  These would be the items recommended to the shopper.  Item 498 is true for the data instance, yet false for the query instance.  Item 498 would be recommended.*

# 5 Rent Prediction  

You have been asked by a San Francisco property investment company to create a predictive model that will generate house price estimates for properties they are considering purchasing as rental properties. The table below lists a sample of properties that have recently been sold for rental in the city. The descriptive features in this dataset are Size (the property size in square feet) and Rent (the estimated monthly rental value of the property in dollars). The target feature, Price, lists the prices that these properties were sold for in dollars.

In [21]:
houses

ID  Size  Rent    Price
0   1  2700  9235  2000000
1   2  1315  1800   820000
2   3  1050  1250   800000
3   4  2200  7000  1750000
4   5  1800  3800  1450000
5   6  1900  4000  1500500
6   7   960   800   720000

  1. Create a k−d tree for this dataset. Assume the following order over the features: Rent, then Size.

In [22]:
tree_seeds = houses[['Size', 'Rent']].to_records(index = False).tolist()
kd_tree = build_treehouse(tree_seeds)

  2. Using the k ⁣− ⁣d tree that you created in the first part of this question, find the nearest neighbor to the following query: Size = 1,000, Rent = 2,200.

In [23]:
# Query for nearest neighbors
n = k_neighbor(kd_tree, [1000, 2200])

print("Nearest Neighbor:", n.item)

Nearest Neighbor: (1315, 1800)
